src/librbd/ImageCtx.cc

```diff
@@ -809,6 +811,20 @@ class ThreadPoolSingleton : public ThreadPool {
     }
   }
 
+  void ImageCtx::clear_pending_completions() {
+    Mutex::Locker l(completed_reqs_lock);
+    ldout(cct, 10) << "clear pending AioCompletion: count="
+                   << completed_reqs.size() << dendl;
+
+    for (xlist<AioCompletion*>::iterator it = completed_reqs.begin();
+         !it.end(); ++it) {
+      ldout(cct, 10) << "clear pending AioCompletion: " << *it << dendl;
+      (*it)->set_event_notify(false);
```

Comment:
When the image is closed, all pending AIO is flushed, so all previously submitted event notify AIO completions have completed and added themselves to ImageCtx::completed_reqs.  Now, the AioCompletion destructor removes itself from ImageCtx::completed_reqs (if the event notify flag is still set).  I am guessing that is what you are trying to protect against with this loop. 

My comment is that the xlist clear method invoked below on line 825 is enough to disassociate the xlist item from the xlist, so the remove_myself() call in the destructor is safe to invoke since it's already been removed from the list.